File: Data Structures & Algorithms/clone-graph/submission-4.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""


class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if node is None:
            return None
        def dfs(curr):
            for neigh in curr.neighbors:
                if neigh.val not in graph:
                    graph[neigh.val] = Node(neigh.val)
                    dfs(neigh)
                graph[curr.val].neighbors.append(graph[neigh.val])
        graph = dict()
        graph[node.val] = Node(node.val)
        dfs(node)
        return graph[node.val]

File: Data Structures & Algorithms/generate-parentheses/submission-4.cpp
class Solution {
public:
    vector<string> results;
    vector<string> generateParenthesis(int n) {
        string curr = "(";
        dfs(1, 0, n, curr);
        return results;
    }
    void dfs(int o, int c, const int n, string& curr) {
        if (c > o or ((c == n) and (o < n))) {return;}
        if (o + c == 2 * n) {
            if (o == c) {
                results.push_back(curr);
            }
            return;
        }
        curr += '(';
        dfs(o + 1, c, n, curr);
        curr.pop_back();
        curr += ')';
        dfs(o, c + 1, n, curr);
        curr.pop_back();
        return;
    }
};
